### PR TITLE
Convert clipping variables in rgblight.c to a structure

### PR DESCRIPTION
**Pull request by drashna**
This allows these variables to be called by keyboard code that uses the custom rgb driver.

Right now, any keyboard using a custom rgblight driver cannot access these settings, so any benefit from using them is lost for those boards.  Such as the Ergodox EZ.  This allows them to be externed, so they can be used by those boards. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mtei** [comment]
FYI:
I think it's better to put those variables in a structure.

Like below:
```diff
diff --git a/quantum/rgblight.h b/quantum/rgblight.h
index e3aa098e4..de90319db 100644
--- a/quantum/rgblight.h
+++ b/quantum/rgblight.h
@@ -170,6 +170,14 @@ typedef struct _rgblight_status_t {
 #    endif
 } rgblight_status_t;
 
+typedef struct _rgblight_ranges_t {
+    uint8_t clipping_start_pos;
+    uint8_t clipping_num_leds;
+    uint8_t effect_start_pos;
+    uint8_t effect_end_pos;
+    uint8_t effect_num_leds;
+} rgblight_ranges_t;
+
 /* === Utility Functions ===*/
```

```diff
diff --git a/quantum/rgblight.c b/quantum/rgblight.c
index 7949bb688..bd9ea6814 100644
--- a/quantum/rgblight.c
+++ b/quantum/rgblight.c
@@ -95,11 +95,7 @@ LED_TYPE led[RGBLED_NUM];
 #    define LED_ARRAY led
 #endif
 
-static uint8_t clipping_start_pos = 0;
-static uint8_t clipping_num_leds  = RGBLED_NUM;
-static uint8_t effect_start_pos   = 0;
-static uint8_t effect_end_pos     = RGBLED_NUM;
-static uint8_t effect_num_leds    = RGBLED_NUM;
+rgblight_ranges_t ranges = { 0, RGBLED_NUM,  0, RGBLED_NUM, RGBLED_NUM };
```

**Comment by drashna** [comment]
It's a lot more changes that I would like to make here, but I think you're right.  It would make things easier, in general. for sure.

**Comment by mtei** [comment]
FYI:

I think there are several ways.

Export rgblight_ranges directly:
```c
/* File:rgblight.h */
extern rgblight_ranges_t rgblight_ranges;
```

Prepare a function that returns a pointer to rgblight_ranges:
```c
/* File:rgblight.h */
extern rgblight_ranges_t *rgblight_get_ranges(void);

/* File:rgblight.c */
rgblight_ranges_t *rgblight_get_ranges(void) {
  return &rgblight_ranges;
}
```

Prepare a new variant of rgblight_set():
```c
/* File:rgblight.c */

__attribute__((weak))
void rgblight_set_with_range(rgblight_ranges_t *range) {
   ...
}

__attribute__((weak))
void rgblight_set(void) {
   rgblight_set_with_range(&rgblight_ranges);
}
```

**Comment by drashna** [comment]
@mtei I think, for now, I'm going to leave this as it is.  Maybe clean up the formatting and such.   And leave the rest to somebody like you, that's a lot more familiar with this code.  

The changes that you propose sound good, but ... I'm not confident enough in my skill and comprehension for much more than this.

**Comment by mtei** [comment]
I'm not sure why RGBLIGHT_CUSTOM_DRIVER was needed.

For example, isn't it possible to override only the output to LEDs as shown below?

```diff
diff --git a/quantum/rgblight.c b/quantum/rgblight.c
index 7949bb688..322982cb2 100644
--- a/quantum/rgblight.c
+++ b/quantum/rgblight.c
@@ -596,6 +596,11 @@ void rgblight_sethsv_master(uint8_t hue, uint8_t sat, uint8_t val) { rgblight_se
 void rgblight_sethsv_slave(uint8_t hue, uint8_t sat, uint8_t val) { rgblight_sethsv_range(hue, sat, val, (uint8_t)RGBLED_NUM / 2, (uint8_t)RGBLED_NUM); }
 #endif  // ifndef RGBLIGHT_SPLIT
 
+__attribute__((weak))
+void rgblight_setrawleds(LED_TYPE *start_led, uint16_t num_leds) {
+    ws2812_setleds(start_led, num_leds);
+}
+
 #ifndef RGBLIGHT_CUSTOM_DRIVER
 void rgblight_set(void) {
     LED_TYPE *start_led;
@@ -620,7 +625,7 @@ void rgblight_set(void) {
 #    else
     start_led = led + clipping_start_pos;
 #    endif
-    ws2812_setleds(start_led, num_leds);
+    rgblight_setrawleds(start_led, num_leds);
 }
 #endif
```

File: quantum/rgblight.h [hunk 1]
```diff
@@ -170,6 +170,21 @@ typedef struct _rgblight_status_t {
...
+ */
+typedef struct _rgblight_ranges_t {
+    uint8_t clipping_start_pos;
+    uint8_t clipping_num_leds;
+    uint8_t effect_start_pos;
+    uint8_t effect_end_pos;
+    uint8_t effect_num_leds;
+} rgblight_ranges_t;
+
+#    ifdef RGBLIGHT_CUSTOM_DRIVER
```
Comment (tzarc) [inline quantum/rgblight.h]: Lets remove the `#ifdef`?

**Comment by mtei** [comment]
@drashna 
I tried to make a sample patch without using RGBLIGHT_CUSTOM_DRIVER. Does this do what you want?

14bbe50296c25eb726e12ebb76aa663e2aa0ee27
9e7f65cada8de87db9565ee5806edabe984d1870

**Comment by spidey3** [comment]
I did some testing on this, and it does seem to be working well. 
It looks like there are a few keyboards where work is needed to make them compile after this refactoring. Is there any way that I can help?

**Comment by drashna** [comment]
@spidey3 if you wanted to do the work for the other boards, you can check out my branch, and open a PR to it.  That would/should give you partial/co-author credit for the PR.

**Comment by spidey3** [comment]
> @spidey3 if you wanted to do the work for the other boards, you can check out my branch, and open a PR to it. That would/should give you partial/co-author credit for the PR.

Sure, I can do that. 
What's the best git-fu to create that checkout?

**Comment by drashna** [comment]
The simple way should be: 
```
git fetch origin refs/pull/7720/head
git checkout -b drashna/rgblight/clipping_range FETCH_HEAD
```
The more complicated way is using hub, or adding my repo as a remote, check out a new branch based on mine.

**Comment by drashna** [comment]
Figured out the issue with the compiling.  

Really, I think that `rgblight_call_driver` removes the need for a custom driver, completely.  The problem is, when you enable the custom driver, the `ws2812_setleds` in the `rgblight_call_driver` function is erroring out.

The fix is either moving the call into the preprocessor with rgblight_set, or just include `ws2812.h` regardless if the custom driver is enabled or not.

**Comment by spidey3** [comment]
> Figured out the issue with the compiling. (@spidey3 heads up)

Thanks!

> Really, I think that `rgblight_call_driver` removes the need for a custom driver, completely...

I am inclined to agree with this. But looking at some of the code out there, at least a few folks are doing custom things that don't really fit comfortably into the purpose of `rgblight_call_driver()` in their custom implementations of `rgblight_set()`.  I again wonder whether there should be a weakly defined `rgblight_set_user()` that they can override, to do those things...

**Comment by drashna** [comment]
Depending on what they're doing, `rgblight_call_driver` passes on a LED array, allows further modification of that array and for to be handled properly.  

For instance, the Ergodox EZ is a great example of this, since it sends the LED info over i2c to half of LEDs, and can be mirrored, duplicated or fully addressible.

**Pull request merged**